ports/esp32/esp32_rmtrx.c

```diff
@@ -0,0 +1,311 @@
+
+    self->rx_config.signal_range_min_ns = min_ns;
+    self->rx_config.signal_range_max_ns = max_ns;
+
+    self->pin = pin_id;
+
+    rmt_rx_channel_config_t rx_ch_conf = {
+        .gpio_num = self->pin = pin_id,
+        .clk_src = RMT_CLK_SRC_DEFAULT,
+        .resolution_hz = resolution_hz,
+        .mem_block_symbols = num_symbols,
+    };
+
+    check_esp_err(rmt_new_rx_channel(&rx_ch_conf, &self->channel));
+    rmt_rx_event_callbacks_t cbs = {
+        .on_recv_done = rmt_recv_done,
+    };
+    check_esp_err(rmt_rx_register_event_callbacks(self->channel, &cbs, self));
+    check_esp_err(rmt_enable(self->channel));
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+static void esp32_rmtrx_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    esp32_rmtrx_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    mp_printf(print, "RMT RX pin=%u min_ns=%u max_ns=%u",
+        self->pin, self->rx_config.signal_range_min_ns, self->rx_config.signal_range_max_ns);
+}
+
+
+static mp_obj_t esp32_rmtrx_release(mp_obj_t self_in) {
+    esp32_rmtrx_obj_t *self = MP_OBJ_TO_PTR(self_in);
+
+    if (self->pin != -1) {
+        self->pin = -1;
+
+        rmt_disable(self->channel);
+        rmt_del_channel(self->channel);
+
+        m_free(self->items);
+        self->cap_items = 0;
+        m_free(self->recv_data);
+        self->recv_data = 0;
+        self->recv_count = 0;
+        self->rx_active = false;
+
+        return mp_const_true;
+    }
+
+    return mp_const_false;
+}
+static MP_DEFINE_CONST_FUN_OBJ_1(esp32_rmtrx_release_obj, esp32_rmtrx_release);
+
+
+static mp_obj_t esp32_rmtrx_deinit(mp_obj_t self_in) {
+    esp32_rmtrx_release(self_in);
+    return mp_const_none;
+}
+static MP_DEFINE_CONST_FUN_OBJ_1(esp32_rmtrx_deinit_obj, esp32_rmtrx_deinit);
+
+
+static mp_obj_t esp32_rmtrx_stop_read_pulses(mp_obj_t self_in) {
+    esp32_rmtrx_obj_t *self = MP_OBJ_TO_PTR(self_in);
+
+    bool ret = self->rx_active;
+    self->rx_active = false;
+    self->recv_count = 0;
+    return ret ? mp_const_true : mp_const_false;
+}
+static MP_DEFINE_CONST_FUN_OBJ_1(esp32_rmtrx_stop_read_pulses_obj, esp32_rmtrx_stop_read_pulses);
+
+
+static mp_obj_t esp32_rmtrx_read_pulses(mp_obj_t self_in) {
+    esp32_rmtrx_obj_t *self = MP_OBJ_TO_PTR(self_in);
+
+    self->rx_active = true;
+    self->recv_count = 0;
+    check_esp_err(rmt_receive(self->channel, self->items, self->cap_items * sizeof(rmt_symbol_word_t), &self->rx_config));
+    return mp_const_none;
+}
+static MP_DEFINE_CONST_FUN_OBJ_1(esp32_rmtrx_read_pulses_obj, esp32_rmtrx_read_pulses);
+
+
+static mp_obj_t esp32_rmtrx_get_data(mp_obj_t self_in) {
+    esp32_rmtrx_obj_t *self = MP_OBJ_TO_PTR(self_in);
+
+    if (!self->recv_count) {
+        return mp_const_none;
+    }
+
+    mp_obj_t list = mp_obj_new_list(self->recv_count, NULL);
+    mp_obj_list_t *list_in = MP_OBJ_TO_PTR(list);
+
+    for (size_t i = 0; i < self->recv_count; i++) {
+        list_in->items[i] = mp_obj_new_int(self->recv_data[i]);
```

Comment:
```suggestion
        list_in->items[i] = MP_OBJ_NEW_SMALL_INT(self->recv_data[i]);
```

Small performance optimisation. (I think it's safe because although recv_data is an array of ints, the actual max size is 17 bits.)

Reply:
Applied but RX has been removed and will be submitted in a new PR.